apps/rush-lib/src/logic/ProjectTaskSelector.ts

```diff
@@ -66,120 +67,142 @@ export class ProjectTaskSelector {
-        buildCacheConfiguration: this._options.buildCacheConfiguration,
-        commandToRun: commandToRun || '',
-        isIncrementalBuildAllowed: this._options.isIncrementalBuildAllowed,
-        projectChangeAnalyzer: this._projectChangeAnalyzer,
-        phase
-      })
-    );
-
-    const dependencyTasks: Set<string> = new Set();
-    if (phase.dependencies?.self) {
-      for (const dependencyPhaseName of phase.dependencies.self) {
-        const dependencyPhase: IPhase = this._getPhaseByName(dependencyPhaseName);
-        const dependencyTaskName: string = this._registerProjectPhaseTask(
+      const customParameterValues: string[] = this._getCustomParameterValuesForPhase(phase);
+
+      for (const project of projectSelection) {
+        const commandToRun: string | undefined = ProjectTaskSelector.getScriptToRun(
           project,
-          dependencyPhase,
-          taskCollection
+          phase.name,
+          customParameterValues
         );
 
-        dependencyTasks.add(dependencyTaskName);
+        if (commandToRun === undefined && !phase.ignoreMissingScript) {
+          throw new Error(
+            `The project [${project.packageName}] does not define a '${phase.name}' command in the 'scripts' section of its package.json`
+          );
+        }
+        const taskName: string = this._getTaskDisplayName(phase, project);
+
+        const task: Task = new Task(
+          new ProjectTaskRunner({
+            rushProject: project,
+            taskName,
+            rushConfiguration: options.rushConfiguration,
+            buildCacheConfiguration: options.buildCacheConfiguration,
+            commandToRun: commandToRun || '',
+            isIncrementalBuildAllowed: options.isIncrementalBuildAllowed,
+            projectChangeAnalyzer,
+            phase
+          }),
+          TaskStatus.Ready
+        );
+
+        taskByOrdinal.set(phaseOffset + project.index, task);
       }
     }
 
-    if (phase.dependencies?.upstream) {
-      for (const dependencyPhaseName of phase.dependencies.upstream) {
-        const dependencyPhase: IPhase = this._getPhaseByName(dependencyPhaseName);
-        for (const dependencyProject of project.dependencyProjects) {
-          const dependencyTaskName: string = this._registerProjectPhaseTask(
-            dependencyProject,
-            dependencyPhase,
-            taskCollection
-          );
+    /**
+     * Enumerates the declared dependencies of the task in (phase * project) key space
+     * task_ordinal = (project_count * phase_index) + project.index
+     */
+    function* getRawDependencies(taskOrdinal: number): Iterable<number> {
+      // eslint-disable-next-line no-bitwise
+      const projectIndex: number = taskOrdinal % projectCount | 0;
+
+      const {
+        phaseDependencies: { self, upstream }
+      }: // eslint-disable-next-line no-bitwise
+      IPhase = ordinalPhases[(taskOrdinal / projectCount) | 0];
```

Comment:
Looking phases up by index seems far more complicated than looking them up by name. What's the benefit to this change?

Reply:
Switching to strings, since the performance effects are negligible